History refactoring for TS

kewne7768 opened the pull request:
This PR cleans up the history stuff to be more TypeScript-friendly. It also re-adds the full display of diamonds/mythos/particles instead of only showing the exponent (this was done because they didn't fit at some point in the past, but it's now hiding useful info for lategame players).

Submitted as draft because I want to fix the last remaining use of `any` and test it a bit more before having it merged.

KhafraDev commented on src/History.ts:
```diff
@@ -1,51 +1,76 @@
+    | ResetHistoryEntryAscend
+
+// A formatter that allows formatting a string. The string should be in a form parsable by break_infinity.js.
+const formatStringOrNumber = (numOrStr: number | Decimal | string) => {
```
```suggestion
const formatStringOrNumber = (numOrStr: DecimalSource) => {
```

KhafraDev commented on src/History.ts:
```diff
@@ -1,51 +1,76 @@
 
-const conditionalFormatPerSecond = (numOrStr: Decimal | number, data: ResetHistoryAscend | ResetHistoryDate) => {
+// A formatter that, if given a number, allows the data to be divided by the amount of seconds spent.
+const conditionalFormatPerSecond = (numOrStr: number | Decimal | string, data: ResetHistoryEntryBase) => {
```
```suggestion
const conditionalFormatPerSecond = (numOrStr: DecimalSource, data: ResetHistoryEntryBase) => {
```

KhafraDev commented on src/History.ts:
```diff
@@ -228,34 +235,38 @@ const resetHistoryRenderRow = (
         const showing = historyGainsOrder[gainIdx];
-        if (data.hasOwnProperty(showing)) {
+        // TODO: Rework this code so we don't have to any
+        const lolLmao = data as any;
```
this of course, as you mentioned ;)

KhafraDev commented on src/Reset.ts:
```diff
@@ -163,17 +168,77 @@ export const updateTesseractAutoBuyAmount = () => {
-    let historyUse = from !== "enterChallenge" && from !== "leaveChallenge";
+const resetAddHistoryEntry = (input: resetNames, from = 'unknown') => {
+    const offeringsGiven = calculateOfferings(input);
+    const isChallenge = ["enterChallenge", "leaveChallenge"].indexOf(from) !== -1;
```
```suggestion
    const isChallenge = ["enterChallenge", "leaveChallenge"].includes(from);
```

KhafraDev commented on src/Reset.ts:
```diff
@@ -225,10 +290,6 @@ export const reset = (input: resetNames, fast = false, from = 'unknown') => {
 
     if (input === "transcension" || input === "transcensionChallenge" || input == "reincarnation" || input == "reincarnationChallenge"
         || input === "ascension" || input === "ascensionChallenge") {
```
```suggestion
    const types = ['transcension', 'transcensionChallenge', 'reincarnation', 'reincarnationChallenge', 'ascension', 'ascensionChallenge'];
    if (types.includes(input)) {
```

kewne7768 replied:
I left this one alone because I didn't touch the code, but I might as well see if GitHub's "commit suggestion" button works

kewne7768 commented:
V2 of the patchset, fixing the use of `any` and just making it a lot stricter overall. I got a bit carried away once I started to clean things up more, so I changed more than I'd have liked. Also, I removed the `any` in the Player object's `.history` as well while I was at it.

I didn't update the files under `dist/`, not sure if it's a good idea to commit those during development because it seems like merge conflict hell.

Pseudonian commented:
Thank you for doing the history refactoring for the TS rewrite. It works wonderfully and it'll help us expand upon the feature in the future. Merged!
